pkg/web/apps/apps_service.go

```diff
@@ -109,3 +108,162 @@ func (a *appsService) BindingAppByApp(appId, name string) error {
 	// if is deleted so just reactive the existent app
 	return a.repository.UnDeleteAppByAppID(app.AppID)
 }
+
+func (a *appsService) checkDeviceWithThisVersion(sdkInfo *models.Device, hasVersion bool, hasApp bool) (bool, *models.Device, error) {
+	// Has no need to do the queries if has not an app or version in the DB with the SDK info
+	if !hasApp || !hasVersion {
+		return false, nil, nil
+	}
+
+	device, err := a.repository.GetDeviceByVersionAndAppID(sdkInfo.Version, sdkInfo.AppID)
+	if err != nil && err != models.ErrNotFound {
+		return false, nil, nil
+	}
+
+	return true, device, nil
+}
+
+func (a *appsService) checkVersionForThisApp(sdkInfo *models.Device, hasApp bool) (bool, *models.Version, error) {
+	// Has no need to do the queries if has not an app in the DB with the SDK info
+	if !hasApp {
+		return false, nil, nil
+	}
+
+	version, err := a.repository.GetVersionByAppIDAndVersion(sdkInfo.AppID, sdkInfo.Version)
+	if err != nil && err != models.ErrNotFound {
+		return false, nil, nil
+	}
+	return true, version, nil
+}
+
+func (a *appsService) checkAnyVersionOfAppForThisDevice(sdkInfo *models.Device, hasVersion bool, hasApp bool) (bool, *models.Device, error) {
+	// Has no need to do the queries if has not an app or version in the DB with the SDK info
+	if !hasApp || hasVersion {
+		return false, nil, nil
+	}
+
+	device, err := a.repository.GetDeviceByDeviceIDAndAppID(sdkInfo.DeviceID, sdkInfo.AppID)
+	if err != nil && err != models.ErrNotFound {
+		return false, nil, nil
+	}
+	return true, device, nil
+}
+
+func (a *appsService) checkIfHasApp(sdkInfo *models.Device) (bool, *models.App, error) {
+	app, err := a.repository.GetAppByAppID(sdkInfo.AppID)
+
+	if err != nil && err != models.ErrNotFound {
+		return false, nil, nil
+	}
+	return true, app, nil
+}
+
+// Init call made from the SDK
+// Response: Version details
+// If the app, version and or device are not tracked in the database then it will be made
+func (a *appsService) InitClientApp(sdkInfo *models.Device) (*models.Version, error) {
+	hasApp, app, err := a.checkIfHasApp(sdkInfo)
```

Comment:
@davidffrench and @craicoverflow 
Note that the approach adopted here got the data then do the actions according to the behaviours. The next step for this impl would refactor it following the Factory design. Doing something as follows. 

Why?

For we are able to attend and accommodate easy any change required as keep the maintainability. 

```go
type initCallFactoryData struct {
	hasApp bool
	hasVersion bool
	hasAnyOtherVersion bool
	hasDeviceVersion bool
	app    *models.App
	deviceDiffVersion, deviceWithVersion,sdkInfo *models.Device
	version *models.Version
	err error
}

// here get all data required 
func (a *appsService) getData(sdkInfo *models.Device) *initCallFactoryData {
	init := new(initCallFactoryData)
	init.hasApp = false
	init.hasVersion = false
	init.hasAnyOtherVersion = false
	init.hasDeviceVersion = false
	init.sdkInfo = sdkInfo
	init.hasApp, init.app,  init.err = a.checkIfHasApp(sdkInfo)
	if init.err != nil {
		return init
	}
	init.hasVersion, init.version, init.err = a.checkVersionForThisApp(sdkInfo, init.hasApp)
	if init.err != nil {
		return init
	}

	init.hasAnyOtherVersion, init.deviceDiffVersion, init.err = a.checkAnyVersionOfAppForThisDevice(sdkInfo, init.hasVersion, init.hasApp)
	if init.err != nil {
		return init
	}

	init.hasDeviceVersion, init.deviceWithVersion, init.err = a.checkDeviceWithThisVersion(sdkInfo, init.hasVersion, init.hasApp)
	if init.err != nil {
		return init
	}
	return init
}

// Here all actions/scenarios impl
func (a *appsService) doActions(initData initCallFactoryData) (*models.Version, error) {
	//** Launch Scenario **
	// If has a version and has tracked a device using this version then it is the launch scenario
	if initData.hasDeviceVersion {
		if initData.err = a.repository.IncrementVersionTotals(initData.deviceWithVersion.VersionID, false); initData.err != nil {
			return nil, initData.err
		}
	}

	//** New install Scenario **
	//If has a version but has NOT tracked a device using this version then it is the new install
	if initData.hasVersion && !initData.hasDeviceVersion {
		if err := a.repository.InsertDeviceOrUpdateVersionID(initData.sdkInfo); err != nil {
			return nil, err
		}
		a.repository.IncrementVersionTotals(initData.version.ID, true)
	}

	// ** New version and new install Scenario **
	// If has NOT the version then it is a new device install of a new version
	if !initData.hasVersion && !initData.hasDeviceVersion {
		if initData.err = a.handleNewVersionAndInstall(initData.sdkInfo, initData.deviceDiffVersion, initData.hasAnyOtherVersion); initData.err != nil {
			return nil, initData.err
		}
	}

	//** New APP Scenario **
	if !initData.hasVersion && !initData.hasDeviceVersion && !initData.hasApp {
		// Create new App with the data sent froom the SDK
		models.NewAppByDevice(initData.sdkInfo)
		if initData.err = a.repository.CreateApp(models.NewAppByDevice(initData.sdkInfo)); initData.err != nil {
			return nil, initData.err
		}
		if initData.err = a.handleNewVersionAndInstall(initData.sdkInfo, initData.deviceDiffVersion, initData.hasAnyOtherVersion); initData.err != nil {
			return nil, initData.err
		}
	}

	// It is to return just the valid information for the device
	response, err := a.preperResponse(initData.version)
	if err != nil {
		return nil, err
	}
	return response, nil
}

// Init call made from the SDK
// Response: Version details
// If the app, version and or device are not tracked in the database then it will be made
func (a *appsService) InitClientApp(sdkInfo *models.Device) (*models.Version, error) {
	initData := a.getData(sdkInfo)
	if initData.err != nil {
		return nil, initData.err
	}
	return a.doActions(*initData)
}
```

Reply:
Note that it still improvements and refinements, however, I hope that it can clarify the idea/approach adopted as it reasons and help you with.